[Merged by Bors] - feat(combinatorics/simple_graph/degree_sum): degree-sum formula and handshake lemma

kmill opened the pull request:
Adds the theorem that the sum of the degrees of the vertices of a simple graph is twice the number of edges.  Also adds corollaries like the handshake lemma, which is that the number of odd-degree vertices is even.

The corollary `exists_ne_odd_degree_if_exists_odd` is in anticipation of Sperner's lemma.

agusakov committed:
added incidence set definitions and lemmas

agusakov committed:
forgot to give myself credit

agusakov committed:
Update src/combinatorics/simple_graph/basic.lean

Co-authored-by: Bhavik Mehta <bhavikmehta8@gmail.com>

agusakov committed:
Update src/combinatorics/simple_graph/basic.lean

Co-authored-by: Bhavik Mehta <bhavikmehta8@gmail.com>

agusakov committed:
fixed typos in docstrings

agusakov committed:
Merge branch 'simple_graphs3' of https://github.com/leanprover-community/mathlib into simple_graphs3
grabbing commits from github

agusakov committed:
definition names and docstrings corrected

agusakov committed:
fixed formatting

agusakov committed:
made changes

agusakov committed:
Update src/combinatorics/simple_graph/basic.lean

Co-authored-by: Bryan Gin-ge Chen <bryangingechen@gmail.com>

agusakov committed:
Merge branch 'master' of https://github.com/leanprover-community/mathlib into simple_graphs3
catching up

agusakov committed:
Merge branch 'simple_graphs3' of https://github.com/leanprover-community/mathlib into simple_graphs3
implementing changes

agusakov committed:
fixed proof (thanks Kyle)

agusakov committed:
oops

agusakov committed:
Update src/combinatorics/simple_graph/basic.lean

Co-authored-by: Aaron Anderson <65780815+awainverse@users.noreply.github.com>

agusakov committed:
changed lemma name

agusakov committed:
Merge branch 'simple_graphs3' of https://github.com/leanprover-community/mathlib into simple_graphs3
fdsjka

agusakov committed:
golfed proof

agusakov committed:
Update src/combinatorics/simple_graph/basic.lean

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

agusakov committed:
Update src/combinatorics/simple_graph/basic.lean

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

agusakov committed:
Merge branch 'master' of https://github.com/leanprover-community/mathlib into simple_graphs3
catching up to mathlib

agusakov committed:
Merge branch 'simple_graphs3' of https://github.com/leanprover-community/mathlib into simple_graphs3
pulling new commits

agusakov committed:
Merge branch 'master' of https://github.com/leanprover-community/mathlib into simple_graphs3
catching up with mathlib

kmill committed:
Merge branch 'simple_graphs3' of https://github.com/leanprover-community/mathlib into degree-sum

kmill committed:
Added the degree-sum formula for simple graphs and some sorry'd corollaries

kmill committed:
exposed combinatorics a bit more

kmill committed:
nonterminal simp

kmill committed:
Merge branch 'master' of https://github.com/leanprover-community/mathlib into degree-sum

kmill committed:
linter (except for some unused arguments that will be used later)

kmill committed:
filled in some sorries, but still need the handshake lemma itself

kmill committed:
found data.nat.parity which has the necessary decidable instances

kmill committed:
Merge remote-tracking branch 'origin/master' into degree-sum

kmill committed:
Filled in proofs, but still have a sorry in a zmod lemma

kmill committed:
simplified these zmod proofs, but still missing key lemma

kmill committed:
found a way to prove the zmod lemmas, now it's complete

kmill committed:
fixed copyright line (had copied from basic.lean)

kmill commented on src/combinatorics/simple_graph/handshake.lean:
```diff
@@ -0,0 +1,259 @@
+end degree_sum
+
+
+section TODO_move
```
Everything in this section needs to find its proper home (or be replaced by something that's already in mathlib somewhere).

bryangingechen replied:
My guess would be `data.zmod.basic`, but I don't know off the top of my head if you can use `even` and `odd` there. If not, maybe a new file in `data.zmod`?

kmill commented on src/combinatorics/simple_graph/handshake.lean:
```diff
@@ -0,0 +1,259 @@
+
+/-- A `dart` is a directed edge, consisting of a vertex and an edge incident to it -/
+@[reducible]
+def dart := Σ (v : V), G.incidence_set v
```
This is reducible so that it automatically has `fintype` instances when `V` is a fintype.

b-mehta replied:
I think you're better off putting the attribute `@[derive fintype]` rather than making it reducible

kmill replied:
It looks like `@[derive fintype]` doesn't work here because it's not able to tell that it also needs the instances `[decidable_eq V]` and `[decidable_rel G.adj]` to synthesize a `fintype`.  I just went ahead and added the instance by hand (using `apply_instance`) and removed `@[reducible]`.

kmill replied:
I also switched it over to being a structure since some of the code was just boilerplate to treat `dart` as if it were one!

b-mehta commented on src/combinatorics/simple_graph/handshake.lean:
```diff
@@ -0,0 +1,259 @@
+def dart_vert : G.dart → V := sigma.fst
+
+/-- Gets the edge that the dart runs through. -/
+def dart_edge : G.dart → sym2 V := λ d, d.snd
```
Does `sigma.snd` not work here?

kmill replied:
This somehow makes Lean automatically coerce a term of the subtype version of `G.incidence_set v` back to a `sym2 V`.

Maybe this would be clearer?
```lean
def dart_edge : G.dart → sym2 V := λ d, (d.snd : sym2 V)
```

b-mehta commented on src/combinatorics/simple_graph/handshake.lean:
```diff
@@ -0,0 +1,259 @@
+d.snd.property
+
+lemma dart.ext {d₁ d₂ : G.dart}
+  (hv : G.dart_vert d₁ = G.dart_vert d₂) (he : G.dart_edge d₁ = G.dart_edge d₂) :
```
```suggestion
@[ext]
lemma dart.ext {d₁ d₂ : G.dart}
  (hv : G.dart_vert d₁ = G.dart_vert d₂) (he : G.dart_edge d₁ = G.dart_edge d₂) :
```

b-mehta commented on src/combinatorics/simple_graph/handshake.lean:
```diff
@@ -0,0 +1,259 @@
+  exact G.dart_edge_eq_iff d' d,
+end
+
+lemma dart_edge_fiber_card_eq_2 (e : sym2 V) (h : e ∈ G.edge_set) :
```
```suggestion
lemma dart_edge_fiber_card (e : sym2 V) (h : e ∈ G.edge_set) :
```

b-mehta commented on src/combinatorics/simple_graph/handshake.lean:
```diff
@@ -0,0 +1,259 @@
+  exact G.dart_edge_fiber_card_eq_2 e h,
+end
+
+/-- The degree-sum formula. -/
```
I think it's worth adding the name `handshaking lemma` here as well (and possibly in the tags too)

b-mehta commented on src/combinatorics/simple_graph/handshake.lean:
```diff
@@ -0,0 +1,259 @@
+begin
+  change (n : zmod 2) = ((0 : ℕ) : zmod 2) ↔ _,
+  rw zmod.eq_iff_modeq_nat,
+  rw nat.even_iff,
```
```suggestion
  rw [zmod.eq_iff_modeq_nat, nat.even_iff],
```

b-mehta commented on src/combinatorics/simple_graph/handshake.lean:
```diff
@@ -0,0 +1,259 @@
+begin
+  change (n : zmod 2) = ((1 : ℕ) : zmod 2) ↔ _,
+  rw zmod.eq_iff_modeq_nat,
+  rw nat.odd_iff,
```
idem

kmill committed:
b-mehta review

kmill committed:
switched from sigma type to structure, removing boilerplate and adding clarity

kmill committed:
linting

kmill committed:
removed unnecessary lemma

eric-wieser commented on src/combinatorics/simple_graph/basic.lean:
```diff
@@ -140,6 +140,12 @@ end
 instance edges_fintype [decidable_eq V] [fintype V] [decidable_rel G.adj] :
   fintype G.edge_set := by { dunfold edge_set, exact subtype.fintype _ }
 
+instance mem_edges_decidable [decidable_rel G.adj] (e : sym2 V) :
```
```suggestion
instance mem_edge_set_decidable [decidable_rel G.adj] (e : sym2 V) :
```

eric-wieser commented on src/combinatorics/simple_graph/handshake.lean:
```diff
@@ -0,0 +1,251 @@
+  apply dart.ext,
+  apply sym2.other_invol',
+  simp only [dart_edge_of_dart_rev],
+end
```
It would be good to provide the `involutive` object too:
```suggestion
@[simp] lemma dart_rev_invol [decidable_eq V] (d : G.dart) : d.rev.rev = d :=
begin
  apply dart.ext,
  apply sym2.other_invol',
  simp only [dart_edge_of_dart_rev],
end

@[simp] lemma dart_rev_involutive [decidable_eq V] : function.involutive (dart.rev : G.dart → G.dart) :=
dart_rev_invol 
```

eric-wieser commented on src/combinatorics/simple_graph/handshake.lean:
```diff
@@ -0,0 +1,251 @@
+  edge_mem := d.edge_mem,
+  vert_mem := sym2.mem_other_mem' d.vert_mem }
+
+@[simp] lemma dart_edge_of_dart_rev [decidable_eq V] (d : G.dart) : d.rev.edge = d.edge := rfl
```
Or maybe:
```suggestion
@[simp] lemma dart.rev_edge [decidable_eq V] (d : G.dart) : d.rev.edge = d.edge := rfl
```

Dot notation might be desirable for a lot of these lemmas, as it means I can do `rw [d.rev_edge]` instead of `rw [dart_edge_of_dart_rev d]`

kmill replied:
Good idea, done.

kmill committed:
unnecessary universe variable

kmill committed:
Update src/combinatorics/simple_graph/basic.lean

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

kmill committed:
Update src/combinatorics/simple_graph/handshake.lean

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

kmill committed:
Merge branch 'degree-sum' of https://github.com/leanprover-community/mathlib into degree-sum

kmill committed:
rename lemmas to use namespace

eric-wieser commented on src/combinatorics/simple_graph/handshake.lean:
```diff
@@ -0,0 +1,254 @@
+  change (n : zmod 2) = ((0 : ℕ) : zmod 2) ↔ _,
+  rw [zmod.eq_iff_modeq_nat, nat.even_iff],
+  trivial,
+end
```
I can golf this to
```suggestion
(char_p.cast_eq_zero_iff (zmod 2) 2 n).trans even_iff_two_dvd.symm
```
But I think my mathlib might be out of date

kmill replied:
Worked in mine, so I change it

kmill committed:
eric-wieser golfing

kmill committed:
(lint) Long line

b-mehta commented on src/combinatorics/simple_graph/handshake.lean:
```diff
@@ -0,0 +1,251 @@
+variables {G}
+
+/-- Reverses the orientation of a dart. -/
+def dart.rev [decidable_eq V] (d : G.dart) : G.dart :=
```
```suggestion
@[simps]
def dart.rev [decidable_eq V] (d : G.dart) : G.dart :=
```

b-mehta commented on src/combinatorics/simple_graph/handshake.lean:
```diff
@@ -0,0 +1,251 @@
+
+lemma dart.neq_rev [decidable_eq V] (d : G.dart) : d ≠ d.rev :=
+begin
+  intro h, rw dart.ext_iff at h,
```
```suggestion
  intro h, 
  rw dart.ext_iff at h,
```

b-mehta commented on src/combinatorics/simple_graph/handshake.lean:
```diff
@@ -0,0 +1,251 @@
+lemma dart.neq_rev [decidable_eq V] (d : G.dart) : d ≠ d.rev :=
+begin
+  intro h, rw dart.ext_iff at h,
+  dsimp [dart.vert, dart.rev] at h,
```
```suggestion
  dsimp at h,
```

b-mehta commented on src/combinatorics/simple_graph/handshake.lean:
```diff
@@ -0,0 +1,251 @@
+  intro h, rw dart.ext_iff at h,
+  dsimp [dart.vert, dart.rev] at h,
+  apply G.edge_other_ne d.edge_mem d.vert_mem,
+  convert h.1.symm, rw sym2.other_eq_other',
```
```suggestion
  convert h.1.symm, 
  rw sym2.other_eq_other',
```

b-mehta commented on src/combinatorics/simple_graph/handshake.lean:
```diff
@@ -0,0 +1,251 @@
+  have h₂ : d₂.edge = ⟦(d₂.vert, d₂.rev.vert)⟧ := (sym2.mem_other_spec' d₂.vert_mem).symm,
+  split,
+  { rw [dart.ext_iff, dart.ext_iff, dart.rev_edge_eq],
+    intro h, simp only [h, and_true, eq_self_iff_true],
```
```suggestion
    intro h, 
    simp only [h, and_true, eq_self_iff_true],
```

b-mehta commented on src/combinatorics/simple_graph/handshake.lean:
```diff
@@ -0,0 +1,251 @@
+    rw [h₁, h₂, sym2.eq_iff] at h,
+    cases h,
+    { exact or.inl h.1 },
+    { exact or.inr h.1 }, },
```
```suggestion
    apply or.imp and.left and.left h },
```

b-mehta commented on src/combinatorics/simple_graph/handshake.lean:
```diff
@@ -0,0 +1,251 @@
+variables (G)
+
+/-- For a given vertex `v`, the injective map from the incidence set at `v` to the darts there. --/
+def dart_from_incidence_set (v : V) : G.incidence_set v → G.dart :=
```
```suggestion
@[simps]
def dart_from_incidence_set (v : V) : G.incidence_set v → G.dart :=
```

b-mehta commented on src/combinatorics/simple_graph/handshake.lean:
```diff
@@ -0,0 +1,251 @@
+
+lemma dart_from_incidence_set_inj (v : V) : function.injective (G.dart_from_incidence_set v) :=
+by { intros e₁ e₂, dsimp only [dart_from_incidence_set, dart_equiv_sigma],
+     simp only [true_and, equiv.coe_fn_symm_mk, eq_self_iff_true], exact subtype.ext,  }
```
```suggestion
λ e₁ e₂ h, by { injection h with h₁ h₂, exact subtype.ext h₂ }
```

b-mehta commented on src/combinatorics/simple_graph/handshake.lean:
```diff
@@ -0,0 +1,251 @@
+  split,
+  { rintro rfl, use [d.edge, ⟨d.edge_mem, d.vert_mem⟩],
+    apply dart.ext; simp [dart_from_incidence_set, dart.vert, dart.edge], },
+  { rintro ⟨e, he, rfl⟩, simp [dart_from_incidence_set, dart.vert] },
```
```suggestion
  { rintro rfl,
    exact ⟨_, ⟨d.edge_mem, d.vert_mem⟩, dart.ext _ _ rfl rfl⟩ },
  { rintro ⟨e, he, rfl⟩,
    refl },
```

b-mehta commented on src/combinatorics/simple_graph/handshake.lean:
```diff
@@ -0,0 +1,251 @@
+begin
+  convert_to (univ : finset G.dart).card = _,
+  rw @card_eq_sum_card_fiberwise _ _ (by assumption) dart.vert _ univ (λ d h, mem_univ _),
+  simp [G.dart_vert_fiber_card_eq_degree],
```
```suggestion
  simp only [←card_univ, ←G.dart_vert_fiber_card_eq_degree],
  exact card_eq_sum_card_fiberwise (by simp),
```

b-mehta commented on src/combinatorics/simple_graph/handshake.lean:
```diff
@@ -0,0 +1,251 @@
+  ext d',
+  simp only [true_and, mem_filter, mem_insert, mem_univ, mem_singleton],
+  exact dart_edge_eq_iff d' d,
+end
```
```suggestion
finset.ext (λ d', by simpa using dart_edge_eq_iff d' d)
```

b-mehta commented on src/combinatorics/simple_graph/handshake.lean:
```diff
@@ -0,0 +1,251 @@
+lemma dart_edge_fiber_card (e : sym2 V) (h : e ∈ G.edge_set) :
+  (univ.filter (λ (d : G.dart), d.edge = e)).card = 2 :=
+begin
+  refine quotient.ind (λ p h, _) e h, cases p with v w,
```
```suggestion
  refine quotient.ind (λ p h, _) e h, 
  cases p with v w,
```

b-mehta commented on src/combinatorics/simple_graph/handshake.lean:
```diff
@@ -0,0 +1,251 @@
+  let d : G.dart := ⟨v, ⟦(v, w)⟧, h, sym2.mk_has_mem _ _⟩,
+  convert_to _ = finset.card {d, d.rev},
+  { rw [card_insert_of_not_mem, card_singleton], simp [d.neq_rev], },
+  congr, apply G.dart_edge_fiber d,
```
```suggestion
  { rw [card_insert_of_not_mem, card_singleton], 
    simp [d.neq_rev], },
  congr, 
  apply G.dart_edge_fiber d,
```

b-mehta commented on src/combinatorics/simple_graph/handshake.lean:
```diff
@@ -0,0 +1,251 @@
+
+lemma dart_card_eq_twice_card_edges : fintype.card G.dart = 2 * G.edge_finset.card :=
+begin
+  convert_to (univ : finset G.dart).card = _,
```
```suggestion
  rw ← card_univ,
```

b-mehta commented on src/combinatorics/simple_graph/handshake.lean:
```diff
@@ -0,0 +1,251 @@
+    (λ e, (univ.filter (λ (x : G.dart), dart.edge x = e)).card) (λ e, 2) _ rfl,
+  simp [mul_comm],
+  intros e h, rw mem_edge_finset at h,
+  exact G.dart_edge_fiber_card e h,
```
```suggestion
  rw [← mul_comm, sum_const_nat],
  intros e h,
  apply G.dart_edge_fiber_card e,
  rwa ← mem_edge_finset,
```

b-mehta commented on src/combinatorics/simple_graph/handshake.lean:
```diff
@@ -0,0 +1,251 @@
+  convert h,
+  ext v, rw ←zmod_neq_zero_iff_odd, split, intro h, convert h, intro h, convert h,
+  intros v, simp only [true_and, mem_filter, mem_univ, ne.def],
+  rw [zmod_eq_zero_iff_even, zmod_eq_one_iff_odd, nat.odd_iff_not_even, imp_self], trivial,
```
```suggestion
  { simp only [filter_congr_decidable, mul_one, nsmul_eq_mul, sum_const, ne.def] at h,
    rw ←zmod_eq_zero_iff_even,
    convert h,
    ext v,
    rw ←zmod_neq_zero_iff_odd,
    congr' },
  { intros v, 
    simp only [true_and, mem_filter, mem_univ, ne.def],
    rw [zmod_eq_zero_iff_even, zmod_eq_one_iff_odd, nat.odd_iff_not_even, imp_self], 
    trivial }
```

b-mehta replied:
and similarly below, each new tactic should be on its own line

kmill committed:
b-mehta review

kmill committed:
Changed structure for dart to be an ordered pair of adjacent vertices

This simplifies a number of things, and it's more natural anyway,
given the structure we use for simple graphs.

b-mehta commented on src/combinatorics/simple_graph/handshake.lean:
```diff
@@ -0,0 +1,256 @@
+@[simp] lemma dart_rev_involutive : function.involutive (dart.rev : G.dart → G.dart) :=
+dart.rev_rev
+
+lemma dart.neq_rev (d : G.dart) : d ≠ d.rev :=
```
I think this would be more natural the other way round: `d.rev ≠ d`, and I think the name `ne` is used more in mathlib than `neq`.

kmill replied:
I also fixed some other `neq`s, thanks

b-mehta commented on src/combinatorics/simple_graph/handshake.lean:
```diff
@@ -0,0 +1,256 @@
+lemma dart.neq_rev (d : G.dart) : d ≠ d.rev :=
+begin
+  cases d with f s h,
+  simp [dart.rev],
```
non-terminal simp?

kmill committed:
oops, non-terminal simp

kmill committed:
factored out combinatorial part of vert fiber

kmill committed:
naming vert -> fst to match struct

bryangingechen commented on src/combinatorics/simple_graph/handshake.lean:
```diff
@@ -0,0 +1,262 @@
+def dart_from_neighbor_set (v : V) : G.neighbor_set v → G.dart :=
+λ w, ⟨v, w, w.property⟩
+
+lemma dart_from_neighbor_set_inj (v : V) : function.injective (G.dart_from_neighbor_set v) :=
```
[See the last sentence of the naming conventions](https://leanprover-community.github.io/contribute/naming.html#injectivity).
```suggestion
lemma dart_of_neighbor_set_injective (v : V) : function.injective (G.dart_from_neighbor_set v) :=
```

bryangingechen commented on src/combinatorics/simple_graph/handshake.lean:
```diff
@@ -0,0 +1,262 @@
+end TODO_move
+
+/-- The handshaking lemma.  See also `simple_graph.sum_degrees_eq_twice_card_edges`. -/
+theorem card_odd_degree_vertices_is_even [fintype V] :
```
```suggestion
theorem even_card_odd_degree_vertices [fintype V] :
```

bryangingechen commented on src/combinatorics/simple_graph/handshake.lean:
```diff
@@ -0,0 +1,262 @@
+  (v : V) (h : odd (G.degree v)) :
+  odd (univ.filter (λ w, w ≠ v ∧ odd (G.degree w))).card :=
+begin
+  rcases G.card_odd_degree_vertices_is_even with ⟨k, hg⟩,
```
```suggestion
  rcases G.even_card_odd_degree_vertices with ⟨k, hg⟩,
```

bryangingechen commented on src/combinatorics/simple_graph/handshake.lean:
```diff
@@ -0,0 +1,262 @@
+end
+
+lemma exists_ne_odd_degree_if_exists_odd [fintype V]
+  (v : V) (h : odd (G.degree v)) :
```
Maybe this is too long...?
```suggestion
lemma exists_ne_odd_degree_of_exists_odd_degree_vertex [fintype V]
  {v : V} (h : odd (G.degree v)) :
```

kmill replied:
I went with `exists_ne_odd_degree_of_exists_odd_degree`.  Naming theory: write `degree_vertices` if plural, and `degree` if singular?

bryangingechen replied:
That was my second choice, so LGTM!

bryangingechen commented on src/combinatorics/simple_graph/handshake.lean:
```diff
@@ -0,0 +1,262 @@
+end
+
+lemma card_odd_degree_vertices_ne_is_odd [fintype V] [decidable_eq V]
+  (v : V) (h : odd (G.degree v)) :
```
```suggestion
lemma odd_card_odd_degree_vertices_ne [fintype V] [decidable_eq V]
  {v : V} (h : odd (G.degree v)) :
```

bryangingechen commented on src/combinatorics/simple_graph/handshake.lean:
```diff
@@ -0,0 +1,262 @@
+
+section TODO_move
+
+lemma zmod_eq_zero_iff_even (n : ℕ) : (n : zmod 2) = 0 ↔ even n :=
```
I think we usually make arguments implicit if they are on both sides of an `iff`.
```suggestion
lemma zmod_eq_zero_iff_even {n : ℕ} : (n : zmod 2) = 0 ↔ even n :=
```

bryangingechen commented on src/combinatorics/simple_graph/handshake.lean:
```diff
@@ -0,0 +1,262 @@
+lemma zmod_eq_zero_iff_even (n : ℕ) : (n : zmod 2) = 0 ↔ even n :=
+(char_p.cast_eq_zero_iff (zmod 2) 2 n).trans even_iff_two_dvd.symm
+
+lemma zmod_eq_one_iff_odd (n : ℕ) : (n : zmod 2) = 1 ↔ odd n :=
```
```suggestion
lemma zmod_eq_one_iff_odd {n : ℕ} : (n : zmod 2) = 1 ↔ odd n :=
```

bryangingechen commented on src/combinatorics/simple_graph/handshake.lean:
```diff
@@ -0,0 +1,262 @@
+  trivial,
+end
+
+lemma zmod_ne_zero_iff_odd (n : ℕ) : (n : zmod 2) ≠ 0 ↔ odd n :=
```
```suggestion
lemma zmod_ne_zero_iff_odd {n : ℕ} : (n : zmod 2) ≠ 0 ↔ odd n :=
```

bryangingechen commented on src/combinatorics/simple_graph/handshake.lean:
```diff
@@ -0,0 +1,262 @@
+
+/-- For a given vertex `v`, the injective map from the incidence set at `v` to the darts there. --/
+def dart_from_neighbor_set (v : V) : G.neighbor_set v → G.dart :=
+λ w, ⟨v, w, w.property⟩
```
```suggestion
/-- For a given vertex `v`, the injective map from the incidence set at `v` to the darts there. -/
def dart_of_neighbor_set (v : V) (w : G.neighbor_set v) : G.dart :=
⟨v, w, w.property⟩
```

bryangingechen commented on src/combinatorics/simple_graph/handshake.lean:
```diff
@@ -0,0 +1,262 @@
+lemma dart_fst_fiber_card_eq_degree [decidable_eq V] (v : V) :
+  (univ.filter (λ d : G.dart, d.fst = v)).card = G.degree v :=
+begin
+  have hh := card_image_of_injective univ (G.dart_from_neighbor_set_inj v),
```
```suggestion
  have hh := card_image_of_injective univ (G.dart_of_neighbor_set_injective v),
```

bryangingechen commented on src/combinatorics/simple_graph/handshake.lean:
```diff
@@ -0,0 +1,262 @@
+  ∃ (w : V), w ≠ v ∧ odd (G.degree w) :=
+begin
+  haveI : decidable_eq V := by { classical, apply_instance },
+  rcases G.card_odd_degree_vertices_ne_is_odd v h with ⟨k, hg⟩,
```
```suggestion
  rcases G.odd_card_odd_degree_vertices_ne h with ⟨k, hg⟩,
```

bryangingechen commented on src/combinatorics/simple_graph/handshake.lean:
```diff
@@ -0,0 +1,262 @@
+- `simple_graph.card_odd_degree_vertices_is_even` is the handshaking lemma.
+- `simple_graph.card_odd_degree_vertices_ne_is_odd` is that the number of odd-degree
+  vertices different from a given odd-degree vertex is odd.
+- `simple_graph.exists_ne_odd_degree_if_exists_odd` is that the existence of an
```
```suggestion
- `simple_graph.even_card_odd_degree_vertices` is the handshaking lemma.
- `simple_graph.odd_card_odd_degree_vertices_ne` is that the number of odd-degree
  vertices different from a given odd-degree vertex is odd.
- `simple_graph.exists_ne_odd_degree_of_exists_odd_degree_vertex` is that the existence of an
```

bryangingechen reviewed (commented):
I think the filename could be a little more general and make contact with the names of its declarations, maybe `degree_sum.lean`?

I made a few name suggestions; apologies if I missed some of the renamings.

LGTM once the TODO_move section is put in the right spot.

kmill committed:
bryangingechen naming suggestions

kmill committed:
renamed handshake.lean to degree_sum.lean

kmill committed:
moved parity lemmas

kmill committed:
It turns out there are two definitions of prime, so it broke

eric-wieser commented on src/combinatorics/simple_graph/degree_sum.lean:
```diff
@@ -0,0 +1,244 @@
+
+/-- There is an equivalence between darts and pairs of a vertex and an incident edge. -/
+@[simps]
+def dart_equiv_sigma : G.dart ≃ Σ v, G.neighbor_set v :=
```
Perhaps
```suggestion
def dart.to_sigma : G.dart ≃ Σ v, G.neighbor_set v :=
```

b-mehta replied:
I think `equiv` is the right name here, the statement is for an equivalence not a function to sigma.

eric-wieser replied:
Counter-example @b-mehta: [`linear_map.to_matrix`](https://leanprover-community.github.io/mathlib_docs/linear_algebra/matrix.html#linear_map.to_matrix) is an equiv, and has type `(M₁ →ₗ[R] M₂) ≃ₗ[R] matrix m n R`. #5237 does the same thing, making `finsupp.to_multiset` an equiv.

In most uses, an equivalence _is_ (used as) a function, so it's convenient to name it after what that function does in many cases.

b-mehta replied:
I think those names are exceptions rather than a pattern, and both cases are equivalences of types with extra structure rather than just an equivalence of types.

eric-wieser replied:
Another example without extra structure is [`multiplicative.to_add`](https://leanprover-community.github.io/mathlib_docs/algebra/group/type_tags.html#multiplicative.to_add)

b-mehta replied:
Perhaps a maintainer can weigh in here? @bryangingechen

bryangingechen replied:
This is a good question. I think recently we have been upgrading lots of functions to `*_equiv`s or `order_iso`s and keeping the original name of the function rather than including `_equiv` or `_order_iso`. 

It's worth getting some more opinions [on Zulip](https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/Naming.20of.20equivs/near/219238479) (and possibly adding something to our naming conventions).

kmill replied:
> In most uses, an equivalence is (used as) a function, so it's convenient to name it after what that function does in many cases.

I think this is one of the few cases where it's not meant to be used as a function -- its only purpose is to establish that `G.dart` is a `fintype` (and it would be nicer if `@[derive fintype]` had worked here).  It's incidentally used as a function for `dart.inhabited`, but that instance probably ought to use `dart_of_neighbor_set` anyway.

kmill committed:
forgot to remove import from data.nat.parity

eric-wieser commented on src/combinatorics/simple_graph/degree_sum.lean:
```diff
@@ -0,0 +1,244 @@
+@[simp] lemma dart.rev_rev (d : G.dart) : d.rev.rev = d :=
+dart.ext _ _ rfl rfl
+
+@[simp] lemma dart_rev_involutive : function.involutive (dart.rev : G.dart → G.dart) :=
```
I wonder if all of these should be in the `dart` namespace. I don't really know what the convention is.

bryangingechen replied:
I considered suggesting this too, but I didn't immediately see any other namespaced `_involutive` declarations in a quick search.

eric-wieser replied:
`list.reverse_involutive` is an easy counter-example. You probably didn't find any because many files would just use `namespace dart` and then put the definitions inside without any explicit `.`s

bryangingechen replied:
Ah, OK. In that case do you think it should be `dart.rev.involutive` or `dart.rev_involutive`? I guess we're not likely to be able to use dot notation with the former.

(I was using the web docs search though which I thought includes full namespaces. However, that search [apparently has other issues](https://github.com/leanprover-community/doc-gen/issues/104).)

eric-wieser replied:
`dart.rev_involutive` seems more consistent with other lemma names.

bryangingechen replied:
```suggestion
@[simp] lemma dart.rev_involutive : function.involutive (dart.rev : G.dart → G.dart) :=
```

eric-wieser commented on src/combinatorics/simple_graph/degree_sum.lean:
```diff
@@ -0,0 +1,244 @@
+
+variables [decidable_eq V]
+
+lemma dart_edge_fiber (d : G.dart) :
```
```suggestion
lemma dart.edge_fiber (d : G.dart) :
```
So that `d.edge_fiber` works

kmill replied:
That's a reasonable suggestion, but I'm torn because the core of the theorem is that `dart_vert_fiber` and `dart_edge_fiber` are meant to identify to fiberings of the `dart` type, and then by summing cardinalities of the fibers you get the degree-sum formula.  I like the parallelism in the names.

It actually should be that `dart_edge_fiber` takes an edge, but because it's a little awkward building the two-element `finset` of darts that map to one, I made it take a dart instead.  Maybe I should look into whether it's so awkward after all.

kmill committed:
and forgot to add that import to data.zmod.parity

kmill committed:
eric-wieser suggestion, also changed inhabited instance to not use equiv

I'm doing this from my laptop, which doesn't have lean installed, so
this might take another commit if I made a mistake.

kmill committed:
added missing brackets (doing this from a plain text editor)

kmill committed:
documentation, some cleanup, some eric-wieser suggestions

bryangingechen reviewed (approved):
Thanks!
bors r+
